### PR TITLE
MRG: Check info consistency

### LINKED ISSUE #2218: bad channels not present
If a channel is not present and I mark it as bad, MNE-Python lets me get away with it. This is not really desired -- makes debugging hard for user.

```py
import mne
from mne import io
from mne.datasets import sample

data_path = sample.data_path()
raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
event_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif'
tmin, tmax = -0.2, 0.5

event_id = {'Auditory/Left': 1, 'Auditory/Right': 2}

raw = io.Raw(raw_fname)
events = mne.read_events(event_fname)

raw.info['bads'] += ['EEG 0723']  # some random bad channel that doesn't exist

# pick EEG and MEG channels
picks = mne.pick_types(raw.info, meg=True, eeg=True, stim=False, eog=True,
                       exclude='bads')
# Read epochs
epochs = mne.Epochs(raw, events, event_id, tmin, tmax, picks=picks,
                    baseline=(None, 0), reject=dict(eeg=80e-6, eog=150e-6),
                    preload=True)
```

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2218.

Added some consistency checking of `info` regarding bads and found a few minor bugs along the way.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
CIs are happy, ready for review/merge from my end.

**Comment by wmvanvliet** [comment]
Sorry, but this is a -1 from me. Calling `_check_info` every time an info object is used is bad design and clutters the code. Moreover, it doesn't solve the problem of allowing inconsistencies in the Info object.
This would do nothing to protect the user from making mistakes in their own code.

An actual solution to the problem would be to make `Info` immutable. Every time you make a change (through an `update()` function or `_setitem__` for example), a new instance of `Info` is created through it's constructor, which is where all sanity checking is performed.

Or don't solve the problem and leave it be. We could just make sure that we use the proper `pick_channels` family of functions consistently throughout the MNE code and don't rely on things like  `[info['ch_names'].index(ch) for ch in channel_names_list]`.

**Comment by Eric89GXL** [comment]
See discussion of the relevant issue, we can't practically make `Info` immutable because of how users have been using it. I think this is the best option we have available. Yes it clutters the code a bit and is not 100% DRY, but it should help avoid errors (and even exposed some in our code). To me those tradeoffs are worth the extra bit of clutter.

**Comment by Eric89GXL** [comment]
I should say that if you can flesh out the `immutable` idea a bit and take care of the examples mentioned in the issue above am open to the idea, but I don't think we can do it in practice. A competing PR would be the best way to demonstrate that such an approach works for our code base and examples without breaking anything...

**Comment by Eric89GXL** [comment]
And FWIW while this won't prevent users from making a mistake in the first place (i.e., adding a bad channel that doesn't exist), it should prevent it from hurting them because as soon as it would matter / be used, they should get an error. So in that sense it is a step forward from before, too.

File: examples/decoding/plot_decoding_csp_eeg.py [hunk 1]
```diff
@@ -57,6 +57,8 @@
 
 # strip channel names
 raw.info['ch_names'] = [chn.strip('.') for chn in raw.info['ch_names']]
+for ch in raw.info['chs']:
+    ch['ch_name'] = ch['ch_name'].strip('.')
```
Comment (agramfort) [inline examples/decoding/plot_decoding_csp_eeg.py]: how about allowing mapping in raw.rename_channels to take a callable as param?
That would avoid exposing the internals of info in an example.

**Comment by agramfort** [comment]
@wmvanvliet I am open to alternative options as @Eric89GXL but this already fixes quite some bugs as shown by the diff

File: examples/decoding/plot_decoding_csp_eeg.py (hunk 1 above)
Comment (Eric89GXL) [inline examples/decoding/plot_decoding_csp_eeg.py]: I haven't looked at that function at all, but it sounds reasonable -- I'll see if I can get it to work

**Comment by Eric89GXL** [comment]
@agramfort done with the last couple of commits

**Comment by Eric89GXL** [comment]
I included a couple of minor docstring fixes I noticed we needed

**Comment by Eric89GXL** [comment]
@wmvanvliet actually `pick_channels` can't be used because it doesn't take `info` as an argument :(

**Comment by Eric89GXL** [comment]
Pushed another commit with some unification/completion of `pick_info`, which wasn't quite complete before (only dealt with `comps` in the raw-writing function, nowhere else).

File: mne/io/meas_info.py [hunk 2]
```diff
@@ -206,6 +206,19 @@ def _anonymize(self):
...
+        if len(missing) > 0:
+            raise RuntimeError('bad channel(s) %s marked do not exist in info'
+                               % (missing,))
+        chs = [ch['ch_name'] for ch in self['chs']]
+        if len(self['ch_names']) != len(chs) or any(
+                ch_1 != ch_2 for ch_1, ch_2 in zip(self['ch_names'], chs)) or \
+                self['nchan'] != len(chs):
+            raise RuntimeError('info channel name inconsistency detected, '
+                               'please notify mne-python developers')
+
```
Comment (agramfort) [inline mne/io/meas_info.py]: add a check that a sfreq is a float as well as low_pass and high_pass if not None

**Comment by agramfort** [comment]
besides +1 for merge

**Comment by wmvanvliet** [comment]
> I think the idea is that people shouldn't ever manually mess with their info structures, other than via the public functions.

I agree that it should not be necessary to mess with the info dictionary directly, mostly because its such a PITA to do so.

**Comment by Eric89GXL** [comment]
@agramfort done

**Comment by agramfort** [comment]
thanks @Eric89GXL

**Pull request merged**